File: spec/java/src/io/kaitai/struct/specwrite/TestBytesPadTerm.java
package io.kaitai.struct.specwrite;

import io.kaitai.struct.ConsistencyError;
import io.kaitai.struct.KaitaiStruct.ReadWrite;
import io.kaitai.struct.testwrite.BytesPadTerm;
import org.testng.annotations.Test;

public class TestBytesPadTerm extends CommonSpec {
    @Test
    public void testCheckGoodMaxLengths() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("12345678901234567890".getBytes());
        r.setStrTerm("12345678901234567890".getBytes());
        r.setStrTermAndPad("12345678901234567890".getBytes());
        r.setStrTermInclude("12345678901234567890".getBytes());
        r._check();
    }

    @Test
    public void testCheckGoodMinLengths() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("".getBytes());
        r.setStrTerm("".getBytes());
        r.setStrTermAndPad("".getBytes());
        r.setStrTermInclude("@".getBytes());
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_pad,.*")
    public void testCheckLongerStrPad() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("123456789012345678901".getBytes());
        r._check();
    }

    @Test
    public void testCheckGoodLastByteStrPad() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad(("@@@@@"+"@@@@@"+"@@@@@"+"@@@@?").getBytes());
        r.setStrTerm("12345678901234567890".getBytes());
        r.setStrTermAndPad("12345678901234567890".getBytes());
        r.setStrTermInclude("12345678901234567890".getBytes());
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_pad,.*")
    public void testCheckBadLastByteStrPad() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("123456789012345678@".getBytes());
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term,.*")
    public void testCheckLongerStrTerm() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("12345678901234567890".getBytes());
        r.setStrTerm("123456789012345678901".getBytes());
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term,.*")
    public void testCheckBadHasTerminator1StrTerm() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("12345678901234567890".getBytes());
        r.setStrTerm("123456789012@4567890".getBytes());
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term,.*")
    public void testCheckBadHasTerminator2StrTerm() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("12345678901234567890".getBytes());
        r.setStrTerm("1234567890123456789@".getBytes());
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term_and_pad,.*")
    public void testCheckLongerStrTermAndPad() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("12345678901234567890".getBytes());
        r.setStrTerm("12345678901234567890".getBytes());
        r.setStrTermAndPad("123456789012345678901".getBytes());
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term_and_pad,.*")
    public void testCheckBadHasTerminatorStrTermAndPad() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("12345678901234567890".getBytes());
        r.setStrTerm("12345678901234567890".getBytes());
        r.setStrTermAndPad("@2345678901234567890".getBytes());
        r._check();
    }

    @Test
    public void testCheckGoodLastByte1StrTermAndPad() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("12345678901234567890".getBytes());
        r.setStrTerm("12345678901234567890".getBytes());
        r.setStrTermAndPad(("+++++"+"+++++"+"+++++"+"++++").getBytes());
        r.setStrTermInclude("12345678901234567890".getBytes());
        r._check();
    }

    @Test
    public void testCheckGoodLastByte2StrTermAndPad() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("12345678901234567890".getBytes());
        r.setStrTerm("12345678901234567890".getBytes());
        r.setStrTermAndPad(("+++++"+"+++++"+"+++++"+"++++0").getBytes());
        r.setStrTermInclude("12345678901234567890".getBytes());
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term_and_pad,.*")
    public void testCheckBadLastByteStrTermAndPad() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("12345678901234567890".getBytes());
        r.setStrTerm("12345678901234567890".getBytes());
        r.setStrTermAndPad("1234567890123456789+".getBytes());
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term_include,.*")
    public void testCheckLongerStrTermInclude() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("12345678901234567890".getBytes());
        r.setStrTerm("12345678901234567890".getBytes());
        r.setStrTermAndPad("12345678901234567890".getBytes());
        r.setStrTermInclude("123456789012345678901".getBytes());
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term_include,.*")
    public void testCheckEmptyStrTermInclude() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("12345678901234567890".getBytes());
        r.setStrTerm("12345678901234567890".getBytes());
        r.setStrTermAndPad("12345678901234567890".getBytes());
        r.setStrTermInclude("".getBytes());
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term_include,.*")
    public void testCheckBadNoTerminatorStrTermInclude() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("12345678901234567890".getBytes());
        r.setStrTerm("12345678901234567890".getBytes());
        r.setStrTermAndPad("12345678901234567890".getBytes());
        r.setStrTermInclude("123".getBytes());
        r._check();
    }

    @Test
    public void testCheckGoodTerminator1StrTermInclude() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("12345678901234567890".getBytes());
        r.setStrTerm("12345678901234567890".getBytes());
        r.setStrTermAndPad("12345678901234567890".getBytes());
        r.setStrTermInclude("12@".getBytes());
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term_include,.*")
    public void testCheckEarlyTerminator1StrTermInclude() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("12345678901234567890".getBytes());
        r.setStrTerm("12345678901234567890".getBytes());
        r.setStrTermAndPad("12345678901234567890".getBytes());
        r.setStrTermInclude("1@@".getBytes());
        r._check();
    }

    @Test
    public void testCheckGoodTerminator2StrTermInclude() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("12345678901234567890".getBytes());
        r.setStrTerm("12345678901234567890".getBytes());
        r.setStrTermAndPad("12345678901234567890".getBytes());
        r.setStrTermInclude("1234567890123456789@".getBytes());
        r._check();
    }

    @Test(expectedExceptions = ConsistencyError.class, expectedExceptionsMessageRegExp = "Check failed: str_term_include,.*")
    public void testCheckEarlyTerminator2StrTermInclude() throws Exception {
        BytesPadTerm r = new BytesPadTerm();
        r.setStrPad("12345678901234567890".getBytes());
        r.setStrTerm("12345678901234567890".getBytes());
        r.setStrTermAndPad("12345678901234567890".getBytes());
        r.setStrTermInclude("123456789012345678@@".getBytes());
        r._check();
    }

    @Override
    protected Class<? extends ReadWrite> getStructClass() {
        return BytesPadTerm.class;
    }

    @Override
    protected String getSrcFilename() {
        return "str_pad_term.bin";
    }
}

File: spec/python/specwrite/test_bytes_pad_term.py
import unittest
from kaitaistruct import ConsistencyError
from specwrite.common_spec import CommonSpec

from testwrite.bytes_pad_term import BytesPadTerm

class TestBytesPadTerm(CommonSpec.Base):
    def __init__(self, *args, **kwargs):
        super(TestBytesPadTerm, self).__init__(*args, **kwargs)
        self.struct_class = BytesPadTerm
        self.src_filename = 'src/str_pad_term.bin'

    def test_check_good_max_lengths(self):
        r = BytesPadTerm()
        r.str_pad = b"12345678901234567890"
        r.str_term = b"12345678901234567890"
        r.str_term_and_pad = b"12345678901234567890"
        r.str_term_include = b"12345678901234567890"
        r._check()

    def test_check_good_min_lengths(self):
        r = BytesPadTerm()
        r.str_pad = b""
        r.str_term = b""
        r.str_term_and_pad = b""
        r.str_term_include = b"@"
        r._check()

    def test_check_longer_str_pad(self):
        r = BytesPadTerm()
        r.str_pad = b"123456789012345678901"
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_pad,"):
            r._check()

    def test_check_good_last_byte_str_pad(self):
        r = BytesPadTerm()
        r.str_pad = b"@@@@@" b"@@@@@" b"@@@@@" b"@@@@?"
        r.str_term = b"12345678901234567890"
        r.str_term_and_pad = b"12345678901234567890"
        r.str_term_include = b"12345678901234567890"
        r._check()

    def test_check_bad_last_byte_str_pad(self):
        r = BytesPadTerm()
        r.str_pad = b"123456789012345678@"
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_pad,"):
            r._check()

    def test_check_longer_str_term(self):
        r = BytesPadTerm()
        r.str_pad = b"12345678901234567890"
        r.str_term = b"123456789012345678901"
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term,"):
            r._check()

    def test_check_bad_has_terminator1_str_term(self):
        r = BytesPadTerm()
        r.str_pad = b"12345678901234567890"
        r.str_term = b"123456789012@4567890"
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term,"):
            r._check()

    def test_check_bad_has_terminator2_str_term(self):
        r = BytesPadTerm()
        r.str_pad = b"12345678901234567890"
        r.str_term = b"1234567890123456789@"
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term,"):
            r._check()

    def test_check_longer_str_term_and_pad(self):
        r = BytesPadTerm()
        r.str_pad = b"12345678901234567890"
        r.str_term = b"12345678901234567890"
        r.str_term_and_pad = b"123456789012345678901"
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term_and_pad,"):
            r._check()

    def test_check_bad_has_terminator_str_term_and_pad(self):
        r = BytesPadTerm()
        r.str_pad = b"12345678901234567890"
        r.str_term = b"12345678901234567890"
        r.str_term_and_pad = b"@2345678901234567890"
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term_and_pad,"):
            r._check()

    def test_check_good_last_byte1_str_term_and_pad(self):
        r = BytesPadTerm()
        r.str_pad = b"12345678901234567890"
        r.str_term = b"12345678901234567890"
        r.str_term_and_pad = b"+++++" b"+++++" b"+++++" b"++++"
        r.str_term_include = b"12345678901234567890"
        r._check()

    def test_check_good_last_byte2_str_term_and_pad(self):
        r = BytesPadTerm()
        r.str_pad = b"12345678901234567890"
        r.str_term = b"12345678901234567890"
        r.str_term_and_pad = b"+++++" b"+++++" b"+++++" b"++++0"
        r.str_term_include = b"12345678901234567890"
        r._check()

    def test_check_bad_last_byte_str_term_and_pad(self):
        r = BytesPadTerm()
        r.str_pad = b"12345678901234567890"
        r.str_term = b"12345678901234567890"
        r.str_term_and_pad = b"1234567890123456789+"
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term_and_pad,"):
            r._check()

    def test_check_longer_str_term_include(self):
        r = BytesPadTerm()
        r.str_pad = b"12345678901234567890"
        r.str_term = b"12345678901234567890"
        r.str_term_and_pad = b"12345678901234567890"
        r.str_term_include = b"123456789012345678901"
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term_include,"):
            r._check()

    def test_check_empty_str_term_include(self):
        r = BytesPadTerm()
        r.str_pad = b"12345678901234567890"
        r.str_term = b"12345678901234567890"
        r.str_term_and_pad = b"12345678901234567890"
        r.str_term_include = b""
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term_include,"):
            r._check()

    def test_check_bad_no_terminator_str_term_include(self):
        r = BytesPadTerm()
        r.str_pad = b"12345678901234567890"
        r.str_term = b"12345678901234567890"
        r.str_term_and_pad = b"12345678901234567890"
        r.str_term_include = b"123"
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term_include,"):
            r._check()

    def test_check_good_terminator1_str_term_include(self):
        r = BytesPadTerm()
        r.str_pad = b"12345678901234567890"
        r.str_term = b"12345678901234567890"
        r.str_term_and_pad = b"12345678901234567890"
        r.str_term_include = b"12@"
        r._check()

    def test_check_early_terminator1_str_term_include(self):
        r = BytesPadTerm()
        r.str_pad = b"12345678901234567890"
        r.str_term = b"12345678901234567890"
        r.str_term_and_pad = b"12345678901234567890"
        r.str_term_include = b"1@@"
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term_include,"):
            r._check()

    def test_check_good_terminator2_str_term_include(self):
        r = BytesPadTerm()
        r.str_pad = b"12345678901234567890"
        r.str_term = b"12345678901234567890"
        r.str_term_and_pad = b"12345678901234567890"
        r.str_term_include = b"1234567890123456789@"
        r._check()

    def test_check_early_terminator2_str_term_include(self):
        r = BytesPadTerm()
        r.str_pad = b"12345678901234567890"
        r.str_term = b"12345678901234567890"
        r.str_term_and_pad = b"12345678901234567890"
        r.str_term_include = b"123456789012345678@@"
        with self.assertRaisesRegex(ConsistencyError, u"^Check failed: str_term_include,"):
            r._check()
